### PR TITLE
Add NetworkPolicy to osg-hosted-ce Stable Application

### PR DESCRIPTION
**Pull request by Mansalu**
Adds a generic NetworkPolicy for restricting traffic to ~and from~ the HostedCE. Allows configuration based on a list of IP CIDR ranges.

```
NetworkPolicy:
  Enabled: false
  AllowedCIDRs:
  - 0.0.0.0/0
```

These ranges get mapped to Ingress ~and Egress~ policy in Kubernetes. By default Policy is turned off and open to all IP ranges.

We want to allow sites running the CE to lockdown the application to only the factory hosts and their own cluster environment.

Egress will remain wide open to prevent issues with OSG Operational services.

### CODE REVIEW

**Comment by brianhlin** [comment]
While we're at it, should we limit ingress to 9619, 80 (for Let's Encrypt), and 8080 for the logging sidecar?

**Comment by Mansalu** [comment]
Yes, I think we should.

**Comment by Mansalu** [comment]
The ports are now added to the ingress rule. Would it be possible/reasonable to include the necessary factory/frontend IP ranges to the default configuration?

**Comment by brianhlin** [comment]
Frontends shouldn't be directly contacting CEs so we shouldn't have to worry about their ranges (and this would be pretty difficult to keep up-to-date). I could see having the factory IPs in the values by default (but keeping `NetworkingPolicy` disabled) making sense. Thoughts @jdost321

**Comment by Mansalu** [comment]
Yes, the plan is to leave it disabled by default for all the apps. But for sites that would like to enjoy the extra security, it would be nice if they didn't have to hunt down the range they need. Even narrowing down the factory range to a /24 would be helpful

**Comment by brianhlin** [comment]
The defaults should also include the GRACC endpoint (for reporting accounting) and the central collectors (grid info services) off the top of my head. Though I'm not sure about all of this because then you're tying the SLATE apps pretty closely with the deployment of OSG Operational services, which aren't necessarily transparent or announced.

File: stable/osg-hosted-ce/osg-hosted-ce/templates/policy.yaml
Comment (brianhlin) [inline stable/osg-hosted-ce/osg-hosted-ce/templates/policy.yaml:26]: Is using the same IP ranges between ingress and egress the right thing to do?

Comment (Mansalu) [reply stable/osg-hosted-ce/osg-hosted-ce/templates/policy.yaml:26]: It's kind of just the simplest thing to do.

 We could definitely build a distinction in the chart values between ranges for ingress and ranges for egress, but it does complicate the configuration, and force the user to think about the distinction between traffic that needs to be received by the CE and traffic that needs to be sent by it. 

This requires a little bit of "grid knowledge", but if we set up nice defaults with good README+comments, I think it would be fine to build in the distinction.

What are your thoughts?

**Comment by Mansalu** [comment]
> 
> 
> The defaults should also include the GRACC endpoint (for reporting accounting) and the central collectors (grid info services) off the top of my head. Though I'm not sure about all of this because then you're tying the SLATE apps pretty closely with the deployment of OSG Operational services, which aren't necessarily transparent or announced.

This is important input. We certainly do not want to inadvertently break things like GRACC or other OSG services with this change. I agree we also want to avoid tying things too closely to OSG Operation Services. If it would be possible to whitelist one or a few broad networks, which would allow wiggle room for unannounced changes, I think that is still advantageous. Many sites would prefer to whitelist the entire IP space available to OSG Ops over running a service that is open to the internet.

**Comment by brianhlin** [comment]
I think then the safest thing to do here is to just allow restriction of ingress and grant unfettered egress until someone requests otherwise. Getting jobs (ingress) are a readily obvious failure mode but missing accounting or info service advertisement (egress) is something that can fly under the radar for a while.

File: stable/osg-hosted-ce/osg-hosted-ce/templates/policy.yaml
Comment (brianhlin) [inline stable/osg-hosted-ce/osg-hosted-ce/templates/policy.yaml:24]: Does this work? We just do the following:

```suggestion
  - {}
```

Comment (Mansalu) [reply stable/osg-hosted-ce/osg-hosted-ce/templates/policy.yaml:24]: Both approaches will work

https://kubernetes.io/docs/concepts/services-networking/network-policies/#default-allow-all-egress-traffic

File: stable/osg-hosted-ce/osg-hosted-ce/README.md
Comment (brianhlin) [inline stable/osg-hosted-ce/osg-hosted-ce/README.md:295]: If `0.0.0.0/0` is all ranges, wouldn't this allow unrestricted ingress? Also, we should probably set `Enabled: True` here.

File: stable/osg-hosted-ce/osg-hosted-ce/README.md
Comment (brianhlin) [inline stable/osg-hosted-ce/osg-hosted-ce/README.md:285]: With the previous sentence, the following example should show the configuration that would restrict access between the OSG factories and the CE

@jdost321 @mmascher what're the factory IP ranges?

File: stable/osg-hosted-ce/osg-hosted-ce/README.md
Comment (brianhlin) [inline stable/osg-hosted-ce/osg-hosted-ce/README.md:278]:
```suggestion
Traffic can be restricted to only allow communication between the CE and OSG Factory:
```
We're not restricting egress so the local cluster environment shouldn't factor in here

File: stable/osg-hosted-ce/osg-hosted-ce/README.md
Comment (brianhlin) [inline stable/osg-hosted-ce/osg-hosted-ce/README.md:272]:
```suggestion
A generic NetworkPolicy for restricting traffic to the HostedCE. Allows configuration based on a list of IP CIDR ranges.
```

File: stable/osg-hosted-ce/osg-hosted-ce/README.md
Comment (Mansalu) [reply stable/osg-hosted-ce/osg-hosted-ce/README.md:295]: Yes. The default policy is wide open and disabled. I am amenable to `Enabled: True` being default if that's what we want. We can also remove `0.0.0.0/0` if we have a good default range.

Comment (brianhlin) [reply stable/osg-hosted-ce/osg-hosted-ce/README.md:295]: I don't think we should change enable this by default. Just that the documentation reads as an example of how to enable a network policy so the docs should 1. set `Enabled: True` and 2. remove `0.0.0.0/0` because presumably if they're enabling this, they want to restrict access

**Comment by Mansalu** [comment]
Any word on default IP Range(s)?

**Comment by brianhlin** [comment]
To start, we can use the SDSC range `169.228.38.32/27`. Eventually we'll want to add the Chicago River and UW Madison Tiger cluster ranges.

**Comment by bbockelm** [comment]
Hi all,

I'd be happy to test this out once merged.  Would rather have an initial good set of CIDRs instead of trying to make things perfect.

Brian

**Comment by Mansalu** [comment]
> 
> 
> Hi all,
> 
> I'd be happy to test this out once merged. Would rather have an initial good set of CIDRs instead of trying to make things perfect.
> 
> Brian

I think we are only missing the IP range for UW Madison Tiger. I got both SDSC and Uchicago River in. Are you saying just merge this now so we can test? And we will add the missing range later?

**Comment by bbockelm** [comment]
For now, you can use `128.104.103.0/24` for the Tiger cluster.  That said, I can always add other NetworkPolicies out-of-band if needed to work around a missing CIDR.

**Review by bbockelm — CHANGES_REQUESTED** [review]
We missed something!

Tried applying the policies by hand and everything imploded.  It seems that the policy blocks the pod from communicating with itself (e.g., the schedd talking to the collector).

I think you need something like:
```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: {{ template "osg-hosted-ce.fullname" . }}-internal
  labels:
    app: {{ template "osg-hosted-ce.name" . }}
    chart: {{ template "osg-hosted-ce.chart" . }}
    release: {{ .Release.Name }}
    instance: {{ .Values.Instance | quote }}
spec:
  podSelector:
    matchLabels:
      app: {{ template "osg-hosted-ce.name" . }}
      chart: {{ template "osg-hosted-ce.chart" . }}
      release: {{ .Release.Name }}
      instance: {{ .Values.Instance | quote }}
  policyTypes:
  - Ingress
  ingress:
  - from:
      podSelector:
        matchLabels:
          app: {{ template "osg-hosted-ce.name" . }}
          chart: {{ template "osg-hosted-ce.chart" . }}
          release: {{ .Release.Name }}
          instance: {{ .Values.Instance | quote }}
    ports:
    - protocol: TCP
      port: 9619
```

**Comment by Mansalu** [comment]
@bbockelm Could you try again when you have time? I have added the rule.

**Comment by bbockelm** [comment]
@Mansalu - bad news :(

At least where I'm testing (the Tiger cluster, which uses MetalLB and Calico), services of type `LoadBalancer` setup NAT in such a way that incoming packets to the pod are stripped of their "real" source IP and just get a cluster IP.

Hence, any attempt to add a CIDR to an ingress is going to fail because the CIDR refers to the real IP but the firewall gets the source IP.  In fact, I found this in the k8s docs:

> Cluster ingress and egress mechanisms often require rewriting the source or destination IP of packets. In cases where this happens, it is not defined whether this happens before or after NetworkPolicy processing, and the behavior may be different for different combinations of network plugin, cloud provider, Service implementation, etc.
> 
> In the case of ingress, this means that in some cases you may be able to filter incoming packets based on the actual original source IP, while in other cases, the "source IP" that the NetworkPolicy acts on may be the IP of a LoadBalancer or of the Pod's node, etc.

So ... not so useful.

**Comment by Mansalu** [comment]
Blocked by Metallb.

**Pull request closed without merging**
